### PR TITLE
Kobe/referral

### PR DESCRIPTION
**Pull request by yattias**
New Referral program changes.
- Referring users will receive a 7% commission for contribution made by invited user
- Invited user will receive a flat 50RSC upon joining

Invited user logic will be gated behind profile claiming in v2

### CODE REVIEW

File: src/oauth/serializers.py [hunk 1]
```diff
@@ -1,30 +1,27 @@
-from allauth.socialaccount.models import SocialAccount
-from allauth.account import app_settings
-from allauth.utils import (
-    get_user_model,
-    email_address_exists
-)
-from rest_framework import serializers
+from time import time
```
Comment (yattias) [inline src/oauth/serializers.py:1]: This file is mostly linting. I'll highlight the actual change below

File: src/oauth/serializers.py [hunk 2]
```diff
@@ -212,45 +196,75 @@ def validate(self, attrs, retry=0):
...
-                user = attrs['user']
-                if referral_code and not user.invited_by and referral_user.id != user.id:
+                user = attrs["user"]
+
+                invited_by_flag_not_set = (
+                    referral_code
+                    and not user.invited_by
+                    and referral_user.id != user.id
+                )
+                if invited_by_flag_not_set:
```
Comment (yattias) [inline src/oauth/serializers.py:222]: This part is new.
This seems to be the best place to pay the invited user bonus.

- Tried using the various allauth signals - request object does not include `referral_code`
- Tried using user `post_save` - but problem is that `invited_by` is set after initial create

Next week, the logic will be moved elsewhere gated by profile claim.

File: src/reputation/distributor.py [hunk 3]
```diff
@@ -84,6 +87,42 @@ def distribute(self):
             print(error_message, e)
         return record
 
+    def _record_referral_distribution_if_applicable(self, original_distribution):
```
Comment (yattias) [inline src/reputation/distributor.py:90]: This part is important. Here we pay a commission to the referer

File: src/user/signals.py [hunk 4]
```diff
@@ -150,45 +150,6 @@ def create_action(sender, instance, created, **kwargs):
                     )
             user = instance.created_by
 
-        """
-        If we're creating an action for the first time,
-        check if we've been referred
-        """
-        referral_content_types = [
```
Comment (yattias) [inline src/user/signals.py:157]: This is old referral program logic where we would pay when invited user created their first contribution

File: src/user/views/user_views.py [hunk 5]
```diff
@@ -72,6 +78,47 @@ def get_serializer_class(self):
     def get_serializer_context(self):
         return {"get_subscribed": True, "get_balance": True, "user": self.request.user}
 
+    @action(detail=False, methods=["GET"], permission_classes=[AllowAny])
```
Comment (yattias) [inline src/user/views/user_views.py:81]: Here we get a list of invited users and the amount of rsc earned by referring them

File: src/oauth/serializers.py [hunk 6]
```diff
@@ -212,45 +192,75 @@ def validate(self, attrs, retry=0):
...
+
+                    has_invited_been_paid = (
+                        Distribution.objects.filter(
+                            distribution_type=REFERRAL_PROGRAM[
+                                "INVITED_DISTRIBUTION_TYPE"
+                            ],
+                            recipient_id=user.id,
+                        ).count()
+                        > 0
+                    )
```
Comment (le-sun) [inline src/oauth/serializers.py:230]:
```suggestion

                    has_invited_been_paid = (
                        Distribution.objects.filter(
                            distribution_type=REFERRAL_PROGRAM[
                                "INVITED_DISTRIBUTION_TYPE"
                            ],
                            recipient_id=user.id,
                        ).exists()
                    )
```

File: src/reputation/distributor.py [hunk 7]
```diff
@@ -71,6 +72,8 @@ def reputation(self):
 
     def distribute(self):
         record = self._record_distribution()
+        self._record_referral_distribution_if_applicable(record)
```
Comment (le-sun) [inline src/reputation/distributor.py:75]: We should move this after line 80, so it only distributes if the original distribution is successful.

File: src/reputation/distributor.py (hunk 3 above)
Comment (le-sun) [inline src/reputation/distributor.py:90]: We should wrap this in a try catch too.

**Review by le-sun — CHANGES_REQUESTED** [review]

File: src/user/views/user_views.py [hunk 8]
```diff
@@ -72,6 +78,47 @@ def get_serializer_class(self):
...
+        for invited_user in invited:
+            for dist in earned_distributions:
+                if dist["giver_id"] == invited_user["user"]["id"]:
+                    invited_user["rsc_earned"] = dist["rsc_earned"]
+
+            invited_user["benefits_expire_on"] = invited_user[
+                "created_date"
+            ] + relativedelta(months=+6)
+
+        return Response(invited)
```
Comment (le-sun) [inline src/user/views/user_views.py:120]:
```suggestion
    def get_referred_users(self, request):
        invited = User.objects.all().filter(
            invited_by__id=request.user.id
        ).annotate(
            rsc_earned=Sum('reputation_records'), 
            benefits_expire_on=models.ExpressionWrapper(models.F('created_date') + datetime.timedelta(days=30), 
            output_field=models.DateTimeField())
    )
```
Something like this works better. You'll have to modify the serializer and logic a bit, but it is better than iterating through the queryset. You can also add a filter on the `Sum` function to filter by specific distributions.

**Review by le-sun — CHANGES_REQUESTED** [review]
Fix merge conflicts too.

File: src/reputation/distributor.py (hunk 7 above)
Comment (yattias) [reply src/reputation/distributor.py:75]: good call!

File: src/reputation/distributor.py (hunk 3 above)
Comment (yattias) [reply src/reputation/distributor.py:90]: good call

File: src/user/views/user_views.py [hunk 9]
```diff
@@ -72,6 +78,41 @@ def get_serializer_class(self):
...
+                    Q(
+                        reputation_handed_out__distribution_type="REFERRAL_REFERER_EARNINGS"
+                    )
+                    & Q(reputation_handed_out__recipient_id=request.user)
+                ),
+            ),
+            benefits_expire_on=models.ExpressionWrapper(
+                models.F("created_date")
+                + timedelta(
+                    days=REFERRAL_PROGRAM["ELIGIBLE_TIME_PERIOD_IN_MONTHS"] * 30
```
Comment (yattias) [inline src/user/views/user_views.py:96]: @le-sun I can't seem to be able to do this in months. Like `months=6`. Tried using relativedelta but doesn't work here. any ideas?

File: src/user/serializers.py [hunk 10]
```diff
@@ -457,6 +459,12 @@ def get_author_profile(self, user):
             sentry.log_error(e)
             return {}
 
+    def get_rsc_earned(self, user):
+        return user.rsc_earned
```
Comment (le-sun) [inline src/user/serializers.py:463]:
```suggestion
        rsc_earned = getattr(user, "rsc_earned", None)
        return rsc_earned
```
This ensures that it won't error out

File: src/user/serializers.py [hunk 11]
```diff
@@ -457,6 +459,12 @@ def get_author_profile(self, user):
...
+
+    def get_benefits_expire_on(self, user):
+        return user.benefits_expire_on
```
Comment (le-sun) [inline src/user/serializers.py:466]:
```suggestion
        benefits_expire_on = getattr(user, "benefits_expire_on", None)
        return benefits_expire_on
```
This ensures that it won't error out

**Review by le-sun — CHANGES_REQUESTED** [review]

File: src/user/serializers.py (hunk 11 above)
Comment (yattias) [reply src/user/serializers.py:466]: ah good good

File: src/reputation/views/bounty_view.py [hunk 12]
```diff
@@ -53,21 +53,21 @@ class BountyViewSet(viewsets.ModelViewSet):
...
         return User.objects.get(id=1)
 
     def _get_dao_fee_recipient(self):
-        user = User.objects.filter(email="community@researchhub.com")
+        user = User.objects.filter(email="kobe@researchhub.com")
         if user.exists():
             return user.first()
 
-        user = User.objects.filter(email="bank@researchhub.com")
+        user = User.objects.filter(email="kobe@researchhub.com")
```
Comment (le-sun) [inline src/reputation/views/bounty_view.py:70]: These values need to be reverted

**Review by le-sun — CHANGES_REQUESTED** [review]

**Review by le-sun — APPROVED** [review]
Looks good

**Pull request merged**